mlflow_yarn/yarn_backend.py

```diff
@@ -0,0 +1,139 @@
+        self._mlflow_run_id = mlflow_run_id
+
+    @property
+    def run_id(self):
+        return self._mlflow_run_id
+
+    def wait(self):
+        result = skein_helper.wait_for_finished(self._skein_client, self._skein_app_id)
+        self._skein_client.close()
+        return result
+
+    def cancel(self):
+        self._skein_client.kill_application(self._skein_app_id)
+        self._skein_client.close()
+
+    def get_status(self):
+        app_report = self._skein_client.application_report(self._skein_app_id)
+        return self._translate_to_runstate(app_report.state)
+
+    def _translate_to_runstate(self, app_state):
+        if app_state == "SUCCEEDED":
+            return RunStatus.FINISHED
+        elif app_state == "KILLED":
+            return RunStatus.KILLED
+        elif app_state == "FAILED":
+            return RunStatus.FAILED
+        elif app_state == "UNDEFINED":
+            return RunStatus.RUNNING
+
+        raise ExecutionException("YARN Application {self._skein_app_id}"
+                                 " has invalid status: {app_state}")
+
+
+class YarnProjectBackend(AbstractBackend):
+    """Implementation of AbstractBackend runninG the job on YARN"""
+
+    def run(self, project_uri, entry_point, params,
+            version, backend_config, tracking_uri, experiment_id):
+        _logger.info('using yarn backend')
+        _logger.info(locals())
+        work_dir = fetch_and_validate_project(project_uri, version, entry_point, params)
+        active_run = get_or_create_run(None, project_uri, experiment_id, work_dir, version,
+                                       entry_point, params)
+
+        _logger.info(f"work_dir={work_dir}")
+        project = load_project(work_dir)
+
+        with tempfile.TemporaryDirectory() as tempdir:
+            entry_point_command = project.get_entry_point(entry_point)\
+                .compute_command(params, tempdir)
+
+            _logger.info(f"entry_point_command={entry_point_command}")
+
+            conda_env_name = _get_or_create_conda_env(project.conda_env_path)
+            conda_path = os.path.join(os.environ["MLFLOW_CONDA_HOME"], "envs", conda_env_name)
+            package_path = _pack_conda_env(conda_path)
+
+            additional_files = []
+            for file in os.listdir(work_dir):
+                full_path = os.path.join(work_dir, file)
+                if os.path.isfile(full_path):
+                    additional_files.append(full_path)
+
+            _logger.info(package_path)
+
+            entry_point, args = try_split_cmd(entry_point_command)
+
+            _logger.info(f"args {entry_point} {args}")
+
+            skein_config = skein_config_builder.build(
+                    module_name=entry_point,
+                    args=args,
+                    package_path=package_path,
+                    additional_files=additional_files,
+                    tmp_dir=tempdir)
+
+            client = skein.Client()

```

Comment:
What about to factorize the client at class level.

I see several way to do that:
* static attribute in class
* Or better add a new function that will allocate the YarnProjectBackend (backend_builder in loader.py in mlflow) . In my example I used the default constructor of the classes, but it could be any other method. This method can manage the client and injects it to the class.

Reply:
Added yarn_backend_builder